File: C++/261.graph-valid-tree.cpp
/*
 * @lc app=leetcode id=261 lang=cpp
 *
 * [261] Graph Valid Tree
 *
 * Given n nodes labeled from 0 to n-1 and a list of undirected edges (each edge is a pair of nodes), write a function to check whether these edges make up a valid tree.
 * 
 * Example 1:
 * Input: n = 5, and edges = [[0,1], [0,2], [0,3], [1,4]]
 * Output: true
 * 
 * Example 2:
 * Input: n = 5, and edges = [[0,1], [1,2], [2,3], [1,3], [1,4]]
 * Output: false
 * Note: you can assume that no duplicate edges will appear in edges. Since all edges are undirected, [0,1] is the same as [1,0] and thus will not appear together in edges.
 * 
 * 
 * 
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
public:
    bool validTree(int n, vector<pair<int, int>>& edges) {
        vector<int> roots(n, -1);
        for (auto edge: edges) {
            int p1 = find(roots, edge.first);
            int p2 = find(roots, edge.second);
            if (p1 == p2) return false;
            roots[p1] = p2; 
        }
        return edges.size() == n -1;
    }
    int find(vector<int>& roots, int i) {
        while (roots[i] != -1) i = roots[i];
        return i;
    }
};
// @lc code=end

File: Python/261.graph-valid-tree.py
#
# @lc app=leetcode id=261 lang=python3
#
# [261] Graph Valid Tree
#
# https://leetcode.com/problems/graph-valid-tree/description/
#
# algorithms
# Medium (48.16%)
# Likes:    3266
# Dislikes: 103
# Total Accepted:    409K
# Total Submissions: 846.9K
# Testcase Example:  '5\n[[0,1],[0,2],[0,3],[1,4]]'
#
# You have a graph of n nodes labeled from 0 to n - 1. You are given an integer
# n and a list of edges where edges[i] = [ai, bi] indicates that there is an
# undirected edge between nodes ai and bi in the graph.
# 
# Return true if the edges of the given graph make up a valid tree, and false
# otherwise.
# 
# 
# Example 1:
# 
# 
# Input: n = 5, edges = [[0,1],[0,2],[0,3],[1,4]]
# Output: true
# 
# 
# Example 2:
# 
# 
# Input: n = 5, edges = [[0,1],[1,2],[2,3],[1,3],[1,4]]
# Output: false
# 
# 
# 
# Constraints:
# 
# 
# 1 <= n <= 2000
# 0 <= edges.length <= 5000
# edges[i].length == 2
# 0 <= ai, bi < n
# ai != bi
# There are no self-loops or repeated edges.
# 
# 
#

# @lc code=start
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        root = [i for i in range(n)]
        for e in edges:
            p1 = self.findRoot(root, e[0])
            p2 = self.findRoot(root, e[1])
            if p1 == p2:
                return False
            root[p1] = p2
        return len(edges) == n-1
    
    def findRoot(self, root, i):
        while root[i] != i:
            root[i] = root[root[i]]
            i = root[i]
        return i
# @lc code=end

